File: cpp/src/standard_library/standard_assert_library.cpp
#include "standard_assert_library.hpp"

#include <sstream>
#include <iostream>

#include "../virtual_machine.hpp"
#include "../values/object_value.hpp"
#include "../utils.hpp"
#include "../scope.hpp"

namespace lysithea_vm
{
    std::shared_ptr<const scope> standard_assert_library::library_scope = create_scope();

    std::shared_ptr<scope> standard_assert_library::create_scope()
    {
        auto result = std::make_shared<scope>();

        auto functions = std::make_shared<object_value>();

        functions->data["true"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            auto top = args.get_index(0);
            if (!top.is_true())
            {
                vm.running = false;
                std::cout << "Assert expected true\n";
            }
        });

        functions->data["false"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            auto top = args.get_index(0);
            if (!top.is_false())
            {
                vm.running = false;
                std::cout << "Assert expected false\n";
            }
        });

        functions->data["equals"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            auto expected = args.get_index(0);
            auto actual = args.get_index(1);
            if (expected.compare_to(actual) != 0)
            {
                vm.running = false;
                std::cout << "Assert expected equals:"
                    << "\nExpected: " << expected.to_string()
                    << "\nActual: " << actual.to_string()
                    << "\n";
            }
        });

        functions->data["notEquals"] = value::make_builtin([](virtual_machine &vm, const array_value &args) -> void
        {
            auto expected = args.get_index(0);
            auto actual = args.get_index(1);
            if (expected.compare_to(actual) == 0)
            {
                vm.running = false;
                std::cout << "Assert expected not equals:"
                    << "\nExpected: " << expected.to_string()
                    << "\nActual: " << actual.to_string()
                    << "\n";
            }
        });

        result->try_set_constant("assert", value(functions));

        return result;
    }
}

File: unity/simple-stack-vm-unity/Assets/Scripts/SimpleStackVM/StandardLibrary/StandardAssertLibrary.cs
using System;
using System.Collections.Generic;

namespace SimpleStackVM
{
    public static class StandardAssertLibrary
    {
        #region Fields
        public static readonly IReadOnlyScope Scope = CreateScope();
        #endregion

        #region Methods
        public static Scope CreateScope()
        {
            var result = new Scope();

            var assertFunctions = new Dictionary<string, IValue>
            {
                {"true", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.Get<BoolValue>(0);
                    if (!top.Value)
                    {
                        vm.Running = false;
                        Console.WriteLine(string.Join("\n", vm.CreateStackTrace()));
                        Console.WriteLine($"Assert expected true");
                    }
                })},

                {"false", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.Get<BoolValue>(0);
                    if (top.Value)
                    {
                        vm.Running = false;
                        Console.WriteLine(string.Join("\n", vm.CreateStackTrace()));
                        Console.WriteLine($"Assert expected false");
                    }
                })},

                {"equals", new BuiltinFunctionValue((vm, args) =>
                {
                    var expected = args.Get(0);
                    var actual = args.Get(1);
                    if (expected.CompareTo(actual) != 0)
                    {
                        vm.Running = false;
                        Console.WriteLine(string.Join("\n", vm.CreateStackTrace()));
                        Console.WriteLine($"Assert expected equals:\nExpected: {expected.ToString()}\nActual: {actual.ToString()}");
                    }
                })},

                {"notEquals", new BuiltinFunctionValue((vm, args) =>
                {
                    var expected = args.Get(0);
                    var actual = args.Get(1);
                    if (expected.CompareTo(actual) == 0)
                    {
                        vm.Running = false;
                        Console.WriteLine(string.Join("\n", vm.CreateStackTrace()));
                        Console.WriteLine($"Assert expected not equals:\nActual: {actual.ToString()}");
                    }
                })}
            };

            result.Define("assert", new ObjectValue(assertFunctions));

            return result;
        }
        #endregion
    }
}